Add user-facing fielderator and can resolve interface

mppf opened the pull request:
* Adds new Reflection.chpl standard module for fielderator and
  can-resolve features
* Deprecates only partially build fielderator interface in ChapelBase
* Replaces PRIM_FIELD_VALUE_BY_NAME with PRIM_FIELD_NAME_TO_NUM since
  PRIM_FIELD_NAME_TO_NUM is more generally useful
* Adds user-facing functions with documentation in Reflection.chpl for
  fielderator and can resolve
* Adds tests

TODO:

* user-facing fielderator procs that return lvalues.  I ran into
  some harder to debug problems when trying that so I am putting it off for
  now.
* consider making fielderator procs into generic type methods, like this:

    proc type ?t.someMethod() { ... }
    to define a type method on any type, and

    proc ?t.someMethod() { ... }
    to define a normal method on any type.

Passed full local testing.
Reviewed by @bradcray.

bradcray commented on compiler/resolution/functionResolution.cpp:
```diff
@@ -5596,6 +5597,7 @@ preFold(Expr* expr) {
       }
       call->replace(result);
     } else if (call->isPrimitive(PRIM_FIELD_ID_BY_NUM)) {
+      /* this is really for unions */
```
I don't really understand this comment.  Is it saying that this primitive is only applicable to unions?  (checking the uses, this seems to be the case).  I guess "really" is the part that makes me stumble -- it suggests that there might be other cases in which one might (try to / be able to / want to) apply it.

mppf replied:
It's only used for unions currently. It's not so clear to me why you would use it otherwise. But it *could* be called on say a record field.

bradcray commented on modules/internal/ChapelBase.chpl:
```diff
@@ -1425,23 +1425,6 @@ module ChapelBase {
-    compilerError("Not yet implemented");
-    return __primitive("field value by name", x, name);
-  }
-
```
Oh sheesh, we already had public routines wrapping these primitives, but just hadn't put them into a standard module or documented them?  I'd completely forgotten that apparently...

mppf replied:
I don't think these were ever used... I think you added them in our original pair-program for the fielderator but then I didn't use them in the IO code or the tests for the fielderator.

bradcray commented:
I timed out before getting all the way through this, so will have to finish tomorrow.  I think we may want to have a group discussion about whether Types.chpl is the right/best place for these routines -- I'm worried it's becoming a bit of a kitchen sink and in the specific case of canResolve(), it doesn't seem deeply connected to types.  But I wanted to look at the docs online before weighing in on that and then timed out after firing off the build.

mppf commented:
OK, thanks for looking at it. I don't have a strong opinion for which module these should go into. Types.chpl seemed the most reasonable of the existing modules.

bradcray commented on modules/standard/Types.chpl:
```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
+/* Return the number of fields in a class or record as a param.
+   The count of fields includes types and param fields.
+ */
+proc numFields(type t) param
```
I realize that you're arguably following the already-established mode used by this file, but I think that in the spirit of making the documentation as clear as possible (and supporting good interfaces), we should declare the return types of library routines whenever possible.  I.e., I'd change the return type to 'param : int' (and, if I'd reviewed the original type documentation, would've required 'param : bool' for all the is*Type queries -- I'm not asking you to change the existing code, though).  I think the getField* routines are the main ones that can't do this because the module code can't assert what the return type is?

bradcray replied:
Did we establish some sort of rule of thumb as to when the ":returns:" field should be used and when it shouldn't, or did we leave it up to the author?  If the former, what was it?  If the latter, what's yours?

mppf replied:
OK, I've added the return types when not generic. I didn't do that originally because of the long-standing (but apparently fixed) bug where param functions didn't work if you specified a return type.

I don't know of any rule of thumb for ":returns:" fields - I tend to use them when the function documentation is complicated enough that it seems useful to describe the arguments separately and individually. These seem simple enough not to bother. Of course, I could write the whole thing in a ":returns:"... but that seems less nice to me.

bradcray commented:
Ultimately, it seems like we'd want to support the fielderator queries as static/type methods on 'object' and 'value' respectively so that one could say C.numFields() rather than numFields(C).  This seems aesthetically far more pleasing if we could pull it off for all the routines.

bradcray commented on modules/standard/Types.chpl:
```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
+pragma "no doc"
+proc fieldNumToName(type t, param i:int) param {
+  compilerError("fieldNumToName deprecated. Use getFieldName");
+}
```
Is this here due to the fact that we had the routines in an undocumented form in the internal modules before, just so someone wouldn't use them accidentally?  Would you add a future which calls this routine to the test/deprecated directory so that we remember to pull this overload after 1.13 goes out?

mppf replied:
OK, I've added such a test.

bradcray commented on modules/standard/Types.chpl:
```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
+proc getField(x:?t, param i:int)
+  return __primitive("field value by num", x, i);
+
+// TODO ref return version of getField(x, int)
```
For my own morbid curiosity, is there a short summary of what kinds of problems the ref return version ran into?

mppf replied:
Yes. The ref return intent is somehow causing the value of the param argument i to no longer be known at compile time in the usual way. Around  functionResolution.cpp:5583, in handling PRIM_FIELD_VALUE_BY_NUM, var is NULL. Is there some other way we should be getting the value of a param arg at compile time?

I thought I would put it off since I'm about to change ref return intent handling...

bradcray replied:
Putting this off is definitely fine with me -- I was just curious where the pain point was.

bradcray commented on modules/standard/Types.chpl:
```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
+
+proc fieldValueByNum(x, param i:int) {
+  compilerError("fieldValueByNum deprecated. Use getField");
+}
```
Ditto the previous deprecated routine comment.  
Also: Should this one be labeled no doc as well?

mppf replied:
OK, done.

bradcray commented on modules/standard/Types.chpl:
```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
+ */
+proc getFieldName(type t, param i:int) param
+  return __primitive("field num to name", t, i);
+
```
If i is not in range, what happens?  If something bad, can we make sure i is in 1..numFields here in module code?  If something reasonable, can you add documentation for it?

Related: "should" feels a little passive-aggressive if in fact it must be in that range.

mppf replied:
Added test cases showing the error. Indicated in the documentation it's a compilation error.

bradcray commented on modules/standard/Types.chpl:
```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
+ */
+inline
+proc getField(x:?t, param i:int)
+  return __primitive("field value by num", x, i);
```
See previous comment about out-of-bounds 'i', "should"

bradcray commented on modules/standard/Types.chpl:
```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
+
+/* Get a field index in a class or record.
+   Will not generate a compilation error if the field
+   is not found.
```
Would you either change this to describe what it does rather than doesn't do or else remove the sentence and rely on the :returns: clause below (either is fine with me -- I just don't want to get into the practice of enumerating all the things that a routine might do but doesn't b/c there are lots of them!).

mppf replied:
OK.

bradcray commented on modules/standard/Types.chpl:
```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
+/* Returns true if a function named `fname` taking in no arguments
+   is available in the current scope.
+   */
+proc canResolve(param fname) param
```
Similar comment to the one on fielderators: I think we should declare the return type and the type of fname

mppf replied:
OK

bradcray commented on modules/standard/Types.chpl:
```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
+/* Returns true if a function named `fname` taking the arguments in args
+   is available in the current scope.
+   */
+proc canResolve(param fname, args ...?k) param
```
Since the 'k' is not used, can it be dropped for simplicity and to avoid the implication that it's meaningful (in the documentation)?

mppf replied:
Done.

bradcray commented on modules/standard/Types.chpl:
```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
+  return getFieldIndex(t, s) > 0;
+
+/* Returns true if a function named `fname` taking in no arguments
+   is available in the current scope.
```
In all these descriptions, does "available" mean something specific to you and different than a phrasing like "can be resolved"?  It feels less definitive to me (e.g., "it's available, but you can't call it") but I'm not seeing if that's intentional or not.

mppf replied:
I started with "can be resolved" but that seemed redundant with the name of the method, and I wasn't sure if we could expect users to know what "resolved" means. I could change it to "could be called" perhaps... although that seems to have other problems. What do you think?

bradcray commented on modules/standard/Types.chpl:
```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
+
+/* Returns true if a method on x named `fname` taking in no arguments
+   is available in the current scope.
+   */
```
Is the lack of :arg: documentation on these routines intentional or an oversight?  The completely generic 'x' is what made me note it here, but I think it applies to previous things as well.

Related: Given that we're a call-by-keyword language, we're (unfortunately) sensitive to changes in argument names down the road.  For that reason, I have to keep reminding myself to try and come up with good long-term names for arguments when writing library routines.  Here, I'd be tempted to replace 'x' with something more descriptive now for fear of wanting to do so later.  'receiver' comes to mind, though I've never been particularly wild about that term...

mppf replied:
what about "obj" or "_this" ?

mppf replied:
The lack of :arg: documentation was not an oversight. It seemed redundant to me and I thought it flowed better to describe in one sentence. I can add it if you think it's important.

bradcray replied:
'obj' or 'object' seems pretty good.  I don't like the underscore in '_this'.  I suppose there's even a chance that 'this' might work (we don't reserve it I don't think, do we?) but that seems like it might be confusing.  This conversation makes me think that this is another routine that would, ideally, be great to have be a type method on the object itself (acknowledging that there are challenges there).

bradcray commented on test/param/ferguson/fielderator-user.chpl:
```diff
@@ -0,0 +1,37 @@
+
+assert(getField(r, "x") == 10);
+assert(getField(r, "y") == 20);
+
```
Nice test!

bradcray commented on test/param/ferguson/fielderator-user.chpl:
```diff
@@ -0,0 +1,37 @@
+assert(getField(r, "x") == 100);
+assert(getField(r, "y") == 200);
+
+*/
```
w/out ref versions of these, how do compiler-generated input routines for records work?  Or does going through the primitives directly work, but it falls apart when we wrap them in module-level code?

mppf replied:
ChapelIO.chpl still uses the primitives directly. Something would go wrong if we switched to wrapping in module code.

bradcray commented on test/param/ferguson/tryresolve-user.chpl:
```diff
@@ -0,0 +1,92 @@
+var x:bool;
+
+x = canResolve("hi0");
+assert(x);
```
Does the use of 'x' in this test add any value / work around some issue compared to doing 'assert(canResolve(...))'?

mppf replied:
No, it doesn't matter. That's just how I wrote it. Sometimes it's nice to have different line numbers for the new call itself.

bradcray commented on test/param/ferguson/tryresolve-user.chpl:
```diff
@@ -0,0 +1,92 @@
+
+x = canResolveMethod(obj, "mymethod2", 1, "blarg");
+assert(!x);
+
```
Nice!

bradcray replied:
Does this work with function calls that have default arguments as well?   Varargs?  Is it worth adding cases for such cases?

mppf replied:
I think it could work with default arguments and with varargs, but I havn't tested it yet. (I'll try that). It won't work for pass-by-arg-name though, at least in the current form.

bradcray replied:
Yeah, I saw the mention of lack of pass-by-arg-name/keyword and can definitely understand that.

bradcray commented on test/users/ferguson/fielderator-ref.chpl:
```diff
@@ -31,9 +31,6 @@ proc myproc(type t, ref m) {
   assert(realEqual(__primitive("field value by num", m, 3), 17.23));
-  assert(__primitive("field value by name", m, "x") == 3);
-  assert(__primitive("field value by name", m, "y") == "hi");
-  assert(realEqual(__primitive("field value by name", m, "z"), 17.23));
```
Since we're not really invested in the primitives anymore (i.e., if they changed name/arguments, we wouldn't necessarily care if the user interface kept working), should we rewrite this test and fielderator.chpl to use the public interfaces for additional exposure/coverage?

mppf replied:
I'd like to leave these older tests using the primitives until we can move ChapelIO.chpl over to the public API.

bradcray commented:
Sorry for all the comments -- I tend to turn up the filter a bit for user-facing interfaces and documentation.  I'm done for this pass.

See also the email thread I kicked off about where to place these in the module space.

mppf commented:
Re supporting fielderator as static/type methods:

Couldn't the same be said for all of the other Type methods, e.g. isRecordType ?

The main issue with doing that is that 'value' and 'object' are not generic, so I don't see how the compiler would do generic resolution to fill in the actual type. This results in a failed attempt in the compiler to coerce type R (a record type) into type value, for example.

It seems to me it would be better to support such a thing like this:

    proc type ?t.someMethod() { ... }

to define a type method on any type, and

    proc ?t.someMethod() { ... }

to define a normal method on any type.

mppf commented:
OK, I think I've addressed your feedback. The open questions are:

  * which file should contain these functions?
 * wording for canResolve "available" vs "can be resolved" vs "could be called" etc
 * any advice you have on the 'ref' return error for fielderator
 * strategy for making fielderator type methods, how worth doing now?

bradcray commented:
> Couldn't the same be said for all of the other Type methods, e.g. isRecordType ?

Probably so.  It didn't occur to me when those were being added (and we probably didn't have type methods at that point).

> It seems to me it would be better to support such a thing like this:
> 
> proc type ?t.someMethod() { ... }
> to define a type method on any type, and
> 
> proc ?t.someMethod() { ... }
> to define a normal method on any type.

That would be pretty sweet...

bradcray commented:
> which file should contain these functions?

My read of the email conversation is that nobody would probably object to putting them into modules/standard/Reflection.chpl (and there'd be several fans of this).

> wording for canResolve "available" vs "can be resolved" vs "could be called" etc

Of these, I think I like "could be called" best then "can be resolved" then "available."

> any advice you have on the 'ref' return error for fielderator

No advice at this point and I think taking this step forward is worthwhile even with the limitation.

> strategy for making fielderator type methods, how worth doing now?

I think this can be put off as well.

For these  last two, since this is moving into a module of its own, how about putting some module level documentation listing these as known issues / plans for future improvements so that users can advocate for them and/or be aware that things may evolve further?

bradcray commented:
BTW, I'm putting this in your court because I don't anticipate needing to review it further.  At the same time, feel free to call me back in to look at something if you think it warrants it.
